File: app/assets/javascripts/channels/submit-channel.js
// app/assets/javascripts/channels/submit-channel.js


import { AppDebug }           from '../application/debug';
import { CableChannel }       from '../shared/cable-channel';
import { hexRand }            from '../shared/random';
import { SubmitResponseBase } from '../shared/submit-response';
import {
    SubmitControlRequest,
    SubmitRequest,
} from '../shared/submit-request';


const CHANNEL = 'SubmitChannel';
const DEBUG   = true;

AppDebug.file('channels/submit-channel', CHANNEL, DEBUG);

/**
 * SubmitChannel
 *
 * @extends CableChannel
 */
export class SubmitChannel extends CableChannel {

    static CLASS_NAME     = CHANNEL;
    static DEBUGGING      = DEBUG;

    static CHANNEL_NAME   = CHANNEL;
    static STREAM_ID      = hexRand();
    static DEFAULT_ACTION = 'submission_request';

    // ========================================================================
    // CableChannel overrides
    // ========================================================================

    /**
     * Create a request object from the provided terms then invoke the server
     * method defined in lookup_channel.rb.
     *
     * @note {@link setCallback} is expected to have been called first.
     *
     * @param {*}      data
     * @param {string} [action]
     *
     * @returns {boolean}
     *
     * @see "SubmitChannel#submission_request"
     * @see "SubmitChannel#submission_control"
     */
    request(data, action) {
        if (data instanceof SubmitControlRequest) {
            return super.request(data, 'submission_control');
        } else {
            return super.request(data, action);
        }
    }

    /**
     * _createRequest
     *
     * @param {*} data
     *
     * @returns {SubmitRequest}
     * @protected
     *
     * @see "SubmitChannel#submission_request"
     */
    _createRequest(data) {
        return SubmitRequest.wrap(data);
    }

    /**
     * _createResponse
     *
     * @param {*} data
     *
     * @returns {SubmitResponseBase}
     * @protected
     *
     * @see "SubmitChannel#initial_response"
     * @see "SubmitChannel#final_response"
     * @see "SubmitChannel#step_response"
     * @see "SubmitChannel#control_response"
     */
    _createResponse(data) {
        return SubmitResponseBase.wrap(data);
    }
}

File: app/channels/submit_channel.rb
# app/channels/submit_channel.rb
#
# frozen_string_literal: true
# warn_indent:           true

__loading_begin(__FILE__)

# Bulk upload submission WebSocket channels.
#
class SubmitChannel < ApplicationCable::Channel

  include SubmissionConcern

  # ===========================================================================
  # :section:
  # ===========================================================================

  public

  # @see file:javascripts/channels/submit-channel.js *DEFAULT_ACTION*
  DEFAULT_ACTION = :submission_request

  # ===========================================================================
  # :section: ApplicationCable::Channel overrides
  # ===========================================================================

  protected

  # The channel for the session.
  #
  # @return [String]
  #
  def stream_name
    @stream_name ||= super(DEFAULT_ACTION)
  end

  # Any cleanup needed when channel is unsubscribed.
  #
  # @return [void]
  #
  def unsubscribed
    super
    # TODO: abort any pending requests
  end

  # ===========================================================================
  # :section:
  # ===========================================================================

  public

  # Receive a submission request from the client.
  #
  # @param [Hash{String=>any,nil}] payload
  #
  # @return [void]
  #
  # @see file:javascripts/shared/cable-channel.js    *request()*
  # @see file:javascripts/channels/submit-channel.js *SubmitChannel*
  #
  def submission_request(payload)
    data = stream_recv(payload, meth: __method__)
    data = SubmissionService::BatchSubmitRequest.wrap(data)
    submission_api.make_request(data, channel: self)
  end

  # Receive a submission control request from the client.
  #
  # @param [Hash{String=>any,nil}] payload
  #
  # @return [void]
  #
  # @see file:javascripts/shared/cable-channel.js    *request()*
  # @see file:javascripts/channels/submit-channel.js *SubmitChannel*
  #
  def submission_control(payload)
    data = stream_recv(payload, meth: __method__)
    data = SubmissionService::ControlRequest.wrap(data)
    submission_api.make_request(data, channel: self)
  end

  # ===========================================================================
  # :section: Class methods
  # ===========================================================================

  public

  # Invoked from another thread to push an initial response back to the client.
  #
  # @param [Hash, Array<Class>] data  Payload data
  # @param [Hash]               opt
  #
  # @option opt [Symbol]  :meth       Passed to #stream_send.
  # @option opt [Boolean] :fatal      Passed to #stream_send.
  #
  # @return [void]
  #
  def self.initial_response(data, **opt)
    str_opt = opt.extract!(:meth, :fatal)
    payload = SubmitChannel::InitialResponse.wrap(data, **opt)
    payload.convert_to_data_url! if invalid_payload_size(payload)
    stream_send(payload, meth: __method__, **str_opt, **opt)
  end

  # Invoked from another thread to push acquired data back to the client.
  #
  # @param [Hash] data                Payload data
  # @param [Hash] opt
  #
  # @option opt [Symbol]  :meth       Passed to #stream_send.
  # @option opt [Boolean] :fatal      Passed to #stream_send.
  #
  # @return [void]
  #
  # @see file:javascripts/shared/cable-channel.js    *response()*
  # @see file:javascripts/channels/submit-channel.js *_createResponse()*
  #
  def self.final_response(data, **opt)
    str_opt = opt.extract!(:meth, :fatal)
    payload = SubmitChannel::FinalResponse.wrap(data, **opt)
    payload.convert_to_data_url! if invalid_payload_size(payload)
    stream_send(payload, meth: __method__, **str_opt, **opt)
  end

  # Invoked from another thread to push intermediate data back to the client.
  #
  # @param [Hash] data                Payload data.
  # @param [Hash] opt
  #
  # @option opt [Symbol]  :meth       Passed to #stream_send.
  # @option opt [Boolean] :fatal      Passed to #stream_send.
  #
  # @return [void]
  #
  # @see file:javascripts/shared/cable-channel.js    *response()*
  # @see file:javascripts/channels/submit-channel.js *_createResponse()*
  #
  def self.step_response(data, **opt)
    str_opt = opt.extract!(:meth, :fatal)
    payload = SubmitChannel::StepResponse.wrap(data, **opt)
    payload.convert_to_data_url! if invalid_payload_size(payload)
    stream_send(payload, meth: __method__, **str_opt, **opt)
  end

  # Invoked from another thread to push control status back to the client.
  #
  # @param [Hash] data                Payload data.
  # @param [Hash] opt
  #
  # @option opt [Symbol]  :meth       Passed to #stream_send.
  # @option opt [Boolean] :fatal      Passed to #stream_send.
  #
  # @return [void]
  #
  # @see file:javascripts/shared/cable-channel.js    *response()*
  # @see file:javascripts/channels/submit-channel.js *_createResponse()*
  #
  # @note Currently unused.
  # :nocov:
  def self.control_response(data, **opt)
    str_opt = opt.extract!(:meth, :fatal)
    payload = SubmitChannel::ControlResponse.wrap(data, **opt)
    payload.convert_to_data_url! if invalid_payload_size(payload)
    stream_send(payload, meth: __method__, **str_opt, **opt)
  end
  # :nocov:

end

__loading_end(__FILE__)
